File: apps/api/src/models/guide.py
from enum import Enum
from typing import Literal, Optional, Union

from pydantic import BaseModel, Field

from .objects import CelestialObject
from .stars import Star


class GuideCategory(str, Enum):
    star = "star"
    object = "object"


class UserLevel(str, Enum):
    beginner = "beginner"
    intermediate = "intermediate"
    advanced = "advanced"


class GuideExplainRequest(BaseModel):
    name: str = Field(..., min_length=1, examples=["Betelgeuse"])
    category: Optional[GuideCategory] = Field(default=None, examples=["star"])
    user_level: UserLevel = Field(default=UserLevel.beginner, examples=["beginner"])
    include_scientific_facts: bool = Field(default=True, examples=[True])
    context: Optional[str] = Field(default=None, examples=["User asked about red supergiants."])


class GuideExplainResponse(BaseModel):
    status: Literal["success"] = Field(default="success", examples=["success"])
    object_found: bool = Field(default=True, examples=[True])
    data: Union[Star, CelestialObject] = Field(...)
    explanation: str = Field(..., examples=["Betelgeuse is a red supergiant star in Orion."])
    key_facts: list[str] = Field(
        ...,
        examples=[
            [
                "Located in the constellation Orion",
                "A red supergiant star",
                "One of the brightest stars visible in the night sky",
            ]
        ],
    )


class GuideExplainNotFoundResponse(BaseModel):
    status: Literal["not_found"] = Field(default="not_found", examples=["not_found"])
    object_found: bool = Field(default=False, examples=[False])
    data: None = Field(default=None)
    explanation: str = Field(
        ...,
        examples=["No star or celestial object named 'Unknown Star' was found in the local dataset."],
    )
    key_facts: list[str] = Field(default_factory=list, examples=[[]])

File: apps/web/src/types/guide.ts
export type GuideCategory = "star" | "object";
export type UserLevel = "beginner" | "intermediate" | "advanced";

export interface GuideExplainRequest {
  name: string;
  category?: GuideCategory;
  user_level: UserLevel;
  include_scientific_facts: boolean;
}

export interface StarData {
  id: number;
  name: string;
  category: "star";
  constellation?: string;
  distance_light_years?: number;
  magnitude?: number;
  spectral_type?: string;
  description?: string;
}

export interface ObjectData {
  id: number;
  name: string;
  category: "object";
  object_type?: string;
  constellation?: string;
  distance_light_years?: number;
  description?: string;
}

export type GuideData = StarData | ObjectData;

export interface GuideExplainSuccess {
  status: "success";
  object_found: true;
  data: GuideData;
  explanation: string;
  key_facts: string[];
}

export interface GuideExplainNotFound {
  status: "not_found";
  object_found: false;
  data: null;
  explanation: string;
  key_facts: string[];
}

export type GuideExplainResponse = GuideExplainSuccess | GuideExplainNotFound;
